File: 002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        if not l1:
            return l2
        if not l2:
            return l1

        if l1.val + l2.val < 10:
            l3 = ListNode(l1.val + l2.val)
            l3.next = self.addTwoNumbers(l1.next, l2.next)
        else:
            l3 = ListNode(l1.val + l2.val - 10)
            tmp = ListNode(1)
            tmp.next = None
            l3.next = self.addTwoNumbers(l1.next, self.addTwoNumbers(tmp, l2.next))

        return l3

    # def addTwoNumbers(self, l1, l2):
    #     if not l1:
    #         return l2
    #     if not l2:
    #         return l1;

    #     val1, val2 = [l1.val], [l2.val]
    #     while l1.next:
    #         val1.append(l1.next.val)
    #         l1 = l1.next

    #     while l2.next:
    #         val2.append(l2.next.val)
    #         l2 = l2.next

    #     str1 = ''.join(str(i) for i in val1[::-1])
    #     str2 = ''.join(str(i) for i in val2[::-1])

    #     tmp = str(int(str1) + int(str2))[::-1]
    #     res = ListNode(int(tmp[0]))
    #     run_res = res

    #     for i in range(1, len(tmp)):
    #         run_res.next = ListNode(int(tmp[i]))
    #         run_res = run_res.next

    #     return res

File: 002-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    var result = new ListNode(0),
        node = result,
        m,
        n,
        i,
        j,
        k,
        sum;

    while (l1 || l2) {
        i = node.val || 0;
        j = (l1 && l1.val) || 0;
        k = (l2 && l2.val) || 0;
        sum = i + j + k;

        if (sum >= 10) {
            m = 1;
            n = sum - 10;
        } else {
            m = 0;
            n = sum;
        }

        l1 = l1 && l1.next;
        l2 = l2 && l2.next;

        if (m > 0) {
            node.next = new ListNode(m);
        } else if (l1 || l2) {
            node.next = new ListNode(0);
        }

        node.val = n;
        if (m || l1 || l2) {
            node = node.next;
        }
    }

    return result;
};